File: wielomian.c
#include <stdio.h>
#include <stdlib.h>

void wczytajWielomian(int tab[], int stopien) {
	
	int i = 0;
	
	for(i = stopien; i >= 0; i--){
		
		if(i == 0) {
			
			printf("Wyraz wolny = ");
			
		} else if (i == 1) {
			
			printf("x * ");
			
		} else {
			
			printf("x^%i * ", i);
			
		}
		
		scanf("%i", &tab[i]);
		
	}
}

void wypiszWielomian(int tab[], int stopien) {
	
	int i = 0;
	
	for(i = stopien; i >= 0; i--) {
		
		if(tab[i] != 0) {
		
			if(i == 0) {
				
				printf("%i", tab[i]);
			
			} else if (i == 1) {
				
				if(tab[i] != 1 && tab[i] != -1) {
					
					printf("%i", tab[i]);
					
				} else if (tab[i] == -1) {
					
					printf("-");
					
				}
				
				printf("x");
			
			} else {
			
				if(tab[i] != 1 && tab[i] != -1) {
					
					printf("%i", tab[i]);
					
				} else if (tab[i] == -1) {
					
					printf("-");
					
				}
			
				printf("x^%i",  i);
			
			}
		
		}
		
		if(tab[i-1] > 0 && i > 0) {
			
			printf("+");
			
		}
		
		
	}
	
	
}

int* mnozenieWielomianow(int w1[], int stopien1, int w2[], int stopien2) {
	
	int stopienMax = stopien1 + stopien2;
	int* wynik = (int*)malloc(sizeof(int) * stopienMax + 1);
	int i = 0, j = 0;
	
	for(i = stopienMax; i >= 0; i--) {
		
		wynik[i] = 0;
		
	}
	
	for(i = stopien1; i >= 0; i--) {
		
		for(j = stopien2; j >=0; j--) {
			
			wynik[i+j] = wynik[i+j] + w1[i] * w2[j];
			
		}
		
	}
	
	return wynik;
}

int main() {
	
	printf("Podaj stopien pierwszego wielomianiu: ");
	
	int n1 = 0;
	scanf("%i", &n1);
	
	printf("Podaj stopien drugiego wielomianiu: ");
	
	int n2 = 0;
	scanf("%i", &n2);

	int w1[n1+1], w2[n2+1];
	
	printf("\nPodaj wartosci pierwszego wielomianu.\n");
	
	wczytajWielomian(w1, n1);
	
	printf("\n\nPodaj wartosci drugiego wielomianu.\n");
	
	wczytajWielomian(w2, n2);
	
	int *wynik = mnozenieWielomianow(w1, n1, w2, n2);
	
	printf("\n\nWynik mnozenia:\n");
	
	printf("(");
	wypiszWielomian(w1, n1);
	printf(") * ");
	
	printf("(");
	wypiszWielomian(w2, n2);
	
	printf(") = ");
	
	wypiszWielomian(wynik, n1 + n2);
	
	printf("\n");
	
	return 0;
}

File: Zadania 1, 2/wielomian.rb
#!/usr/bin/env ruby
# Marcin Mackiewicz 215505

def wczytajWielomian(tab)

	(0..tab.length-1).each do |i|

		if tab.length-1-i == 0

			print "Wyraz wolny = "

		elsif tab.length-1-i == 1

			print "x * "

		else

			print "x^#{tab.length-1-i} * "

		end

		tab[tab.length-1-i] = Integer(gets.chomp)

	end

end

def wypiszWielomian(tab)

	(0..tab.length-1).each do |i|

		if tab[tab.length-1-i] != 0

			if tab.length-1-i == 0

				print "#{tab[tab.length-1-i]}"

			elsif tab.length-1-i == 1

				if tab[tab.length-1-i] != 1 && tab[tab.length-1-i] != -1
				print "#{tab[tab.length-1-i]}"

				elsif tab[tab.length-1-i] == -1

					print "-"

				end

				print "x"

			else

				if tab[tab.length-1-i] != 1 && tab[tab.length-1-i] != -1
				print "#{tab[tab.length-1-i]}"

				elsif tab[tab.length-1-i] == -1

					print "-"

				end

				print "x^#{tab.length-1-i}"

			end

		end

		if tab[tab.length-2-i] > 0 && tab.length-1-i != 0

			print "+"

		end

	end

end

def mnozenieWielomianow(w1, w2)

	stopienMax = (w1.length) + (w2.length)

	wynik = Array.new(stopienMax-2)

	(0..wynik.length).each do |i|

		wynik[i] = 0

	end

	(0..w1.length-1).each do |i|

		(0..w2.length-1).each do |j|
	
			wynik[(w1.length-1-i) + (w2.length-1-j)] = wynik[(w1.length-1-i) + (w2.length-1-j)] + (w1[w1.length-1-i] * w2[w2.length-1-j])

		end

	end

	wynik

end

print "Podaj stopien pierwszego wielomianu: "

n1 = Integer(gets.chomp)

w1 = Array.new(n1+1)

print "Podaj stopien drugiego wielomianu: "

n2 = Integer(gets.chomp)

w2 = Array.new(n2+1)

puts "Podaj wartosci pierwszego wielomianu: "

wczytajWielomian(w1)

puts "Podaj wartosci drugiego wielomianu: "

wczytajWielomian(w2)

puts "Wynik mnozenia:"

wynik = mnozenieWielomianow(w1, w2)

print "("
wypiszWielomian(w1)
print ") * "

print "("
wypiszWielomian(w2)
print ") = "

wypiszWielomian(wynik)

puts ""

